### DISCUSSION TITLE
lldb dived into python stack

### DISCUSSION THREAD

**Discussion by j8xixo12** [solvcon/devenv#7] [General (discussions)]
I have studied cpython for a while, I found python interpreter stored many informations into `PyFrameObject`, what I care about is **line number**. 

In order to get python script's line number, I used `co_lontab` which is bytecode offset <----> line number mapping to calculate python source code line number. 

The cpython also provides a helper function `PyCode_Addr2Line(PyCodeObject *co, int addrq)` to convert bytecode offset to line number, and `PyFrame_GetLineNumber(PyFrameObject *f)` also can get line number directly from `PyFrameObject`.

Here is my test python script 
```
def foo1():
    print("Hi")
    return 1 + 1

def foo2(input):
    print(input)
    return

def main():
    a = foo1()
    foo2("1234")
    return

if __name__ == '__main__':
    main()
```

I used `br s -n _PyEval_EvalFrame -c '((int)strcmp(PyUnicode_AsUTF8(f->f_code->co_name), "foo1")) == 0'` to set breakpoint at first line of  `main()`.

Here is the output
```
(lldb) br s -n _PyEval_EvalFrame -c '((int)strcmp(PyUnicode_AsUTF8(f->f_code->co_name), "foo1")) == 0'
Breakpoint 18: 3 locations.
(lldb) r /tmp/test.py
Process 15283 launched: '/Users/laijunxu/.pyenv/versions/3.9-dev-debug/bin/python3.9' (x86_64)
Process 15283 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 18.1
    frame #0: 0x00000001000b9753 python3.9`_PyEval_EvalFrame(tstate=0x0000000100a06460, f=0x0000000100cf45f0, throwflag=0) at pycore_ceval.h:40:12
   37   static inline PyObject*
   38   _PyEval_EvalFrame(PyThreadState *tstate, PyFrameObject *f, int throwflag)
   39   {
-> 40       return tstate->interp->eval_frame(tstate, f, throwflag);
   41   }
   42
   43   extern PyObject *_PyEval_EvalCode(
Target 0: (python3.9) stopped.
(lldb) py-list
    5    def foo2(input):
    6        print(input)
    7        return
    8
    9    def main():
  >10        a = foo1()
   11        foo2("1234")
   12        return
   13
   14    if __name__ == '__main__':
   15        main()
(lldb) p PyFrame_GetLineNumber(f)
(int) $47704 = 10
```
I tried to use `PyFrame_GetLineNumber` to get line number, and I got what I want, but when I tried 
using `br s -n _PyEval_EvalFrame -c '(int)PyFrame_GetLineNumber(f) == 10'` to put breakpoint at the same line
lldb didn't hit the breakpoint.....

I don't know what's going on?

**Comment by yungyuc** [comment]
If `lldb` conditional `br` doesn't work, how about trying to mimic what `pdb` does for (Python) breakpoint?  I assume `pdb` doesn't use either `gdb` nor `lldb`.

**Comment by j8xixo12** [comment]
I plan to write some gdb or lldb script that user can directly put break point into python script.
I wish to finish it before end of Chinese new year.
Hope I can accomplish it as schedule.

**Reply by yungyuc** [reply]
I think getting one of the debugger script to work is good enough as the first step.  Once you have a prototype, we can start to discuss how to make a PR.
